Add ESS Plot

imperorrp opened the pull request:
Adding ESS plot (#5 ) 

Currently implemented for kind='local' only, using the new `scatter_xy` visual element function also added as part of this commit. The ess data ('y' values) obtained from the ess statistical computation via Arviz-Stats is combined with xdata ('x' values) generated via `np.linspace` (like the legacy ess plot in old Arviz does) after this xdata is broadcasted to fit the shape of the ess data. These are concatenated along a new plot_axis dimension (coords 'x' and 'y') which the `scatter_xy` visual element function then splits and plots accordingly. 

### Outputs:

`azp.plot_ess(data)`
![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/0b76930a-ff8f-4523-9091-d58c8a542121)

`azp.plot_ess(data, var_names=["mu", "tau"])`
![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/b7c47960-62fc-4053-9f24-8d4ff4ff8fe9)




<!-- readthedocs-preview arviz-plots start -->
----
📚 Documentation preview 📚: https://arviz-plots--58.org.readthedocs.build/en/58/

<!-- readthedocs-preview arviz-plots end -->

OriolAbril commented on src/arviz_plots/plots/essplot.py:
```diff
@@ -0,0 +1,251 @@
+            print(f"\n xdata_da ={xdata_da}")
+
+            # broadcasting xdata_da to match shape of each variable in ess_y_dataset
+            xdata_broadcasted_dict = {}
+            for var in ess_y_dataset.data_vars:
+                _, xdata_broadcasted = xr.broadcast(ess_y_dataset[var], xdata_da)
+                xdata_broadcasted_dict[var] = xdata_broadcasted
```
```suggestion
            xdata_broadcasted_dict = {var_name: xdata_da.broadcast_like(da) for var_name, da in ess_y_dataset.items()}
```

This should generate the same dictionary being slightly more efficient

could also be merged with the `xdata_dataset` line so it is `xdata_dataset = xr.Dataset({var_name: ...})`

OriolAbril commented on src/arviz_plots/plots/essplot.py:
```diff
@@ -0,0 +1,251 @@
+            ["__variable__"]
+            + [dim for dim in distribution.dims if dim not in {"model"}.union(sample_dims)],
+        )
+        if "model" in distribution:
+            pc_kwargs["aes"] = pc_kwargs.get("aes", {}).copy()
+            pc_kwargs["aes"].setdefault("color", ["model"])
```
```suggestion
        if "model" in distribution:
            pc_kwargs["aes"] = pc_kwargs.get("aes", {}).copy()
            pc_kwargs["aes"].setdefault("color", ["model"])
```

For this plot I think it would look better to also add `x` to the default aesthetic mappings when there are multiple models.

OriolAbril commented on src/arviz_plots/plots/essplot.py:
```diff
@@ -0,0 +1,251 @@
+            print(f"\n xdata_dataset = {xdata_dataset}")
+
+            # concatenating xdata_dataset and ess_y_dataset along plot_axis
+            ess_dataset = xr.concat([xdata_dataset, ess_y_dataset], dim="plot_axis")
+
+            # assigning 'x' and 'y' coordinates to the 'plot_axis' dimension
+            ess_dataset["plot_axis"] = ["x", "y"]
```
```suggestion
            ess_dataset = xr.concat([xdata_dataset, ess_y_dataset], dim="plot_axis").assign_coords(plot_axis=["x", "y"])
```

OriolAbril commented on src/arviz_plots/visuals/__init__.py:
```diff
@@ -71,6 +71,15 @@ def scatter_x(da, target, backend, y=None, **kwargs):
+
+    The input argument `da` is split into x and y using the dimension ``plot_axis``.
+    """
+    plot_backend = import_module(f"arviz_plots.backend.{backend}")
```
not yet, but after merging the plot_ridge PR, this should use `_process_da_x_y` too. See https://github.com/arviz-devs/arviz-plots/pull/57#discussion_r1664857354. The `x` as aesthetic I mentioned above won't work otherwise.

OriolAbril reviewed (commented):
Looks very good so far

imperorrp commented:
Fixed the docstring switch-up for type 'local' and 'quantile' and incorporated other suggestions including the x aesthetic addition for the 'model' dimension. I made a temporary addition to `scatter_xy` to add the extra x arg passed due to this new aesthetic. The output ends up looking like this though- 

```
pc = azp.plot_ess(
        {"centered": data, "non centered": data2}, var_names=["mu", "tau"]
    )
pc.add_legend("model")
```
![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/421926fb-c4b5-4a9a-acc0-f8593d2e6e0b)

The default x arg value generated for the second model is '1',  which makes the data get severely skewed out of the original x axis range. Should we hardcode some smaller 'x' values with something like `pc_kwargs['x'].setdefault(np.linspace(0, 0.01, 9))` maybe?

imperorrp commented:
Updated x aesthetic mapping for multiple-model cases: 

![image](https://github.com/user-attachments/assets/3b2c98ab-a776-421b-b0bd-91a345d8d8a2)

The logic followed is as below. The x_diff is calculated, and currently one-third of that is taken as the range within which points of different models can be plotted. `np.linspace` then ensures an even distribution of the points of whatever number of models to plot within the aforementioned range. 
```
  # setting x aesthetic to np.linspace(-x_diff/3, x_diff/3, length of 'model' dim)
  # x_diff = span of x axis (1) divided by number of points to be plotted (n_points)
  x_diff = 1 / n_points
  if "x" not in pc_kwargs:
      pc_kwargs["x"] = np.linspace(-x_diff / 3, x_diff / 3, distribution.sizes["model"])
  pc_kwargs["aes"].setdefault("x", ["model"])
```

And quantile plots added- 
![image](https://github.com/user-attachments/assets/02e17a24-12b8-4666-805c-873219635933)

Labelling, along the y axes ('ESS for small intervals' for kind='sample', 'ESS for quantiles' for type='quantile') and x axis ('quantile') has also been added

imperorrp commented:
Added rugplot to `plot_ess`. 

Like `plot_trace`, I've set the 'overlay' aesthetic for the 'chain' dimension but is ignored with `.difference()` when the default aes_map is generated for the local and quantile artists. 

Also made a modification to the `trace_rug` visual element to add a new arg `scale`, which helps in controlling the `xvalues` range. It is set to the length of the draw dimension so that the xvalues range is kept between 0-1. 

#### Plot output:

`azp.plot_ess(data, var_names=["mu", "tau"], rug=True)`
![image](https://github.com/user-attachments/assets/56ceeec9-dd87-43b7-9f91-b5fe825739f6)

imperorrp commented:
To do: Apply `rankdata` function to the data once available in Arviz-Stats for proper rug generation.

OriolAbril commented on src/arviz_plots/plots/essplot.py:
```diff
@@ -0,0 +1,438 @@
+            pc_kwargs["aes"].setdefault("color", ["model"])
+            # setting x aesthetic to np.linspace(-x_diff/3, x_diff/3, length of 'model' dim)
+            # x_diff = span of x axis (1) divided by number of points to be plotted (n_points)
+            x_diff = 1 / n_points
+            if "x" not in pc_kwargs:
+                pc_kwargs["x"] = np.linspace(-x_diff / 3, x_diff / 3, distribution.sizes["model"])
```
```suggestion
            n_models = distribution.sizes["model"]
            x_diff = min(1 / n_points / 3, 1 / n_points * n_models / 10)
            pc_kwargs.setdefault("x", np.linspace(-x_diff, x_diff, n_models))
```

Seeing the plot now, 1/3 would probably look good for several models, but for only 2 of them the points end up equispaced (the model1 point corresponding to quantile 0 is as far from the model0 point corresponding to quantile 0 than from the model0 point corresponding to quantile 0.05). This change will result in smaller spacing with small number of models, but never going over 1/3 of the available space.

OriolAbril commented on src/arviz_plots/plots/essplot.py:
```diff
@@ -0,0 +1,438 @@
+                ],
+                dim="ess_dim",
+            )
+            # print(f"\n ess_y_dataset = {ess_y_dataset}")
+
+            # converting xdata into a xr dataarray
+            xdata_da = xr.DataArray(xdata, dims="ess_dim")
+            # print(f"\n xdata_da ={xdata_da}")
+
+            # broadcasting xdata_da to match shape of each variable in ess_y_dataset and
+            # creating a new dataset from dict of broadcasted xdata
+            xdata_dataset = xr.Dataset(
+                {var_name: xdata_da.broadcast_like(da) for var_name, da in ess_y_dataset.items()}
+            )
+            # print(f"\n xdata_dataset = {xdata_dataset}")
+
+            # concatenating xdata_dataset and ess_y_dataset along plot_axis
+            ess_dataset = xr.concat([xdata_dataset, ess_y_dataset], dim="plot_axis").assign_coords(
+                plot_axis=["x", "y"]
+            )
+            print(f"\n distribution = {distribution!r}")
+            print(f"\n ess_dataset = {ess_dataset!r}")
+
+            # step 4
+            # if "color" not in ess_aes:
+            #    ess_kwargs.setdefault("color", "gray")
+
+            # step 5
+            plot_collection.map(
+                scatter_xy, "local", data=ess_dataset, ignore_aes=ess_ignore, **ess_kwargs
+            )
```
all this can be removed

OriolAbril commented on src/arviz_plots/plots/essplot.py:
```diff
@@ -0,0 +1,438 @@
+                ],
+                dim="ess_dim",
+            )
+
+            # converting xdata into an xr datarray
+            xdata_da = xr.DataArray(xdata, dims="ess_dim")
+
+            # broadcasting xdata_da to match shape of each variable in ess_y_dataset and
+            # creating a new dataset from dict of broadcasted xdata
+            xdata_dataset = xr.Dataset(
+                {var_name: xdata_da.broadcast_like(da) for var_name, da in ess_y_dataset.items()}
+            )
+
+            # concatenating xdata_dataset and ess_y_dataset along plot_axis
+            ess_dataset = xr.concat([xdata_dataset, ess_y_dataset], dim="plot_axis").assign_coords(
+                plot_axis=["x", "y"]
+            )
+            print(f"\n ess_dataset = {ess_dataset!r}")
+
+            # step 4
+            # if "color" not in ess_aes:
+            #    ess_kwargs.setdefault("color", "gray")
+
+            # step 5
+            plot_collection.map(
+                scatter_xy, "quantile", data=ess_dataset, ignore_aes=ess_ignore, **ess_kwargs
+            )
```
If all this is moved outside the if. To get the same exact behaviour using `.map(scatter_xy, kind, ...)` would be needed, but from #66 I think it should instead be `.map(scatter_xy, "ess", ...)`

OriolAbril reviewed (commented):
go over plot_ess_evolution review first

imperorrp commented:
Just made updates. This takes into account #66 and plots the mean, sd and min_ess as well using `line_xy`. Users can also set custom text for xlabel and ylabel now:
![image](https://github.com/user-attachments/assets/08ae4602-af60-47b0-a564-b18344e80ca5)
![image](https://github.com/user-attachments/assets/ae3f174b-4d0e-43b3-8bfa-b72915b6b80d)



Annotating these lines is left as well as fixing an issue when the model dimension exists, where the output is like this:
![image](https://github.com/user-attachments/assets/0da6a6e9-0d24-4519-9b1c-010b5e99f173)

OriolAbril commented on src/arviz_plots/plots/essplot.py:
```diff
@@ -0,0 +1,432 @@
+            )
+            print(f"\n mean_ess = {mean_ess}")
+
+            if "linestyle" not in mean_aes:
+                if backend == "matplotlib":
+                    mean_kwargs.setdefault("linestyle", "--")
+                elif backend == "bokeh":
+                    mean_kwargs.setdefault("linestyle", "dashed")
```
you have to get the 2nd linestyle 

```
linestyle = plot_bknd.get_default_aes("linestyle", 2, {})[1]
```

or alternatively, the same way we have defined `|` must be a valid marker for all backends because it is part of the "common plotting interface" we could do the same for solid/dashed linestyles. For now I think I prefer the approach above though

imperorrp replied:
Linestyles -> mean, sd, min
Colors -> for diff models

OriolAbril commented on src/arviz_plots/plots/essplot.py:
```diff
@@ -0,0 +1,432 @@
+            elif backend == "bokeh":
+                min_ess_kwargs.setdefault("linestyle", "dashed")
+
+        plot_collection.map(
+            line_xy,
+            "min_ess",
+            data=min_ess_dataset,
+            ignore_aes=min_ess_ignore,
+            x=x_range,
+            **min_ess_kwargs,
+        )
```
can't you use `x=x_range, y=min_ess` 1d array of length 2 for x and a scalar for y? if both x and y are provided and da doesn't have `plot_axis` (which it doesn't) then `da` is used for looping but ignored for plotting which is what we want here.

OriolAbril commented on src/arviz_plots/plots/essplot.py:
```diff
@@ -0,0 +1,432 @@
+        aes_map = {}
+    else:
+        aes_map = aes_map.copy()
+    aes_map.setdefault(kind, plot_collection.aes_set.difference({"overlay"}))
+    aes_map.setdefault("rug", {"overlay"})
```
sd and mean should get the `color` aesthetic mapped to them, at least when `model` dimension is present. Otherwise you get n_model values of ess_mean for each plot which can't be used to plot a horizontal line at that y coordinate.

Moreover, if color is mapped to the model, then it can be used to distinguish the mean line from the sd line, they should get different linestyles, or there should be different aesthetic mappings defined depending on whether there are multiple models or not.

OriolAbril commented on src/arviz_plots/plots/essplot.py:
```diff
@@ -0,0 +1,432 @@
+
+        min_ess_dataset = xr.Dataset(min_ess_data)
+        print(f"\n min_ess = {min_ess_dataset}")
+
```
this should get "black", "red", or "gray" as default color. Probably gray.

imperorrp commented on src/arviz_plots/plots/essplot.py:
```diff
@@ -0,0 +1,432 @@
+    # overlaying divergences(or other 'rug_kind') for each chain
+    if rug:
+        sample_stats = get_group(dt, "sample_stats", allow_missing=True)
+        rug_kwargs = copy(plot_kwargs.get("rug", {}))
+        if rug_kwargs is False:
+            raise ValueError("plot_kwargs['rug'] can't be False, use rug=False to remove the rug")
```
Adding this `plot_forest` style check and moving it higher up for mcseplot as well

imperorrp commented:
Plot_ess output is now like this:

![image](https://github.com/user-attachments/assets/7bf6a734-5652-4b1f-b915-83c8148153a4)

![image](https://github.com/user-attachments/assets/9cedee7c-a851-474d-9210-03a6bb53809a)

![image](https://github.com/user-attachments/assets/d21e43d5-58f1-47eb-93cf-117735a536ef)


Each of 'mean', 'sd', and 'min_ess' has their own linestyle to distinguish between them. Maybe having another legend for them could be of help though as it might not be evident to a user which is which unless the the linestyle cycle and order of plotting is known by looking at the source code. 

The 'min_ess' line is set to gray by default and 'mean' and 'sd' to the first color in the color cycle by default. If the 'model' dim is present, then all three elements get the same color for a model.

OriolAbril commented:
I think annotating the lines is he only way to clearly indicate which is which. If you set `extra_methods=True` in current arviz, you'll see the output looks like this:

![imatge](https://github.com/user-attachments/assets/6a3eb83b-f6ea-433d-b1fd-83a490000e60)

I would add these two annotations as visual elements here too

imperorrp commented:
> I think annotating the lines is he only way to clearly indicate which is which. If you set `extra_methods=True` in current arviz, you'll see the output looks like this:
> 
> ![imatge](https://private-user-images.githubusercontent.com/23738400/355272274-6a3eb83b-f6ea-433d-b1fd-83a490000e60.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI5NTIyMjksIm5iZiI6MTcyMjk1MTkyOSwicGF0aCI6Ii8yMzczODQwMC8zNTUyNzIyNzQtNmEzZWI4M2ItZjZlYS00MzNkLWIxZmQtODNhNDkwMDAwZTYwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA2VDEzNDUyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEyMzA2YzFjMTdlZmM2N2NhNjY2NTY2NDc5M2Q4NjE2YmRlZGU0MDIzYWY3Zjk3MzU2ODQ5YWIxMTRjY2NiZDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.P6PD2v7v2Xo5Rp0cSerN2oo2_zUhe6TQNPMlsGQvl_A)
> 
> I would add these two annotations as visual elements here too

Todo: Add a new visual element for these annotations

imperorrp commented:
Rebased essplot commits

imperorrp commented:
Added new `annotate_xy` visual element with in-built logic to determine whether to vertically align the intended annotating text top or bottom based on an `extra_da` arg passed to it. This helps when two lines like mean and sd both are being plotted. Defaults for them both are also set- 'bottom' for mean and 'top' for sd. 

New `plot_kwargs` keys `mean_text` and `sd_text` are also added and their default aesthetics from the `aes_map`s for `mean` and `sd` are also used for these if not set. 

#### Outputs:
![image](https://github.com/user-attachments/assets/33777f61-f7ac-4e57-a5d3-78de312475cb)

![image](https://github.com/user-attachments/assets/9a164146-1d7b-4091-941b-711f24a31c85)

imperorrp commented:
- Some examples to the docstring and to the example gallery for plot_ess were added. Documentation seems to build fine when tried locally. 

- Tests were added in `test_plots.py` and `test_hypothesis_plots.py`. In the former, testing with the `datatree_sample` fixture causes the ess computing function from Arviz-Stats to raise an error due to number of sample_dims not being equal to 2, which it seems to require. In the latter, a couple of errors are being raised that have to be looked into. One is with the time limit being exceeded.

imperorrp commented:
Fixed hypothesis test errors for ESS plot

OriolAbril commented on pyproject.toml:
```diff
@@ -27,7 +27,7 @@ classifiers = [
 dependencies = [
   "arviz-base==0.2",
-  "arviz-stats[xarray]==0.2",
+  "arviz-stats[xarray] @ git+https://github.com/arviz-devs/arviz-stats@rankdata",
```
```suggestion
  "arviz-stats[xarray] @ git+https://github.com/arviz-devs/arviz-stats",
```

OriolAbril commented on src/arviz_plots/plots/essplot.py:
```diff
@@ -0,0 +1,550 @@
+            mean_text_kwargs.setdefault("x", 1)
+            mean_text_kwargs.setdefault("horizontal_align", "right")
+
+            # pass the mean vertical_align data for vertical alignment setting
+            if mean_va_align is not None:
+                vertical_align = mean_va_align
+            else:
+                vertical_align = "bottom"
```
also use setdefault

OriolAbril commented on src/arviz_plots/plots/essplot.py:
```diff
@@ -0,0 +1,550 @@
+            sd_text_kwargs.setdefault("x", 1)
+            sd_text_kwargs.setdefault("horizontal_align", "right")
+
+            # pass the sd vertical_align data for vertical alignment setting
+            if sd_va_align is not None:
+                vertical_align = sd_va_align
+            else:
+                vertical_align = "top"
```
also setdefault

OriolAbril commented on src/arviz_plots/visuals/__init__.py:
```diff
@@ -47,6 +47,7 @@ def line(da, target, backend, xname=None, **kwargs):
 
 def trace_rug(da, target, backend, mask, xname=None, y=None, **kwargs):
     """Create a rug plot with the subset of `da` indicated by `mask`."""
+    # print(f'\n da = {da}')
```
delete

OriolAbril commented on src/arviz_plots/visuals/__init__.py:
```diff
@@ -125,6 +138,19 @@ def _ensure_scalar(*args):
+def annotate_xy(da, target, backend, *, text, x=None, y=None, vertical_align=None, **kwargs):
+    """Annotate a point (x, y) in a plot."""
+    if vertical_align is not None:
+        # print(f"\n vertical_align.item() = {vertical_align.item()}")
```
delete

OriolAbril commented on src/arviz_plots/plots/essplot.py:
```diff
@@ -0,0 +1,550 @@
+            pc_kwargs.setdefault("x", np.linspace(-x_diff, x_diff, n_models))
+            pc_kwargs["aes"].setdefault("x", ["model"])
+        aux_dim_list = [dim for dim in pc_kwargs["cols"] if dim != "__variable__"]
+        plot_collection = PlotCollection.wrap(
+            distribution,
+            backend=backend,
+            **pc_kwargs,
+        )
```
add automatic figure sizing, see https://github.com/arviz-devs/arviz-plots/blob/main/src/arviz_plots/plots/forestplot.py#L258 and https://github.com/arviz-devs/arviz-plots/blob/main/src/arviz_plots/plots/traceplot.py#L114

OriolAbril commented on tests/test_hypothesis_plots.py:
```diff
@@ -192,3 +192,73 @@ def test_plot_ridge(datatree, combined, plot_kwargs, labels_shade_label):
+
+@st.composite
+def ess_n_points(draw):
+    return draw(st.integers(min_value=1, max_value=50))  # should this range be changed?
```
range 1-5

OriolAbril commented on tests/test_hypothesis_plots.py:
```diff
@@ -192,3 +192,73 @@ def test_plot_ridge(datatree, combined, plot_kwargs, labels_shade_label):
+        },
+    ),
+    kind=ess_kind_value,
+    relative=ess_relative,
+    rug=ess_rug,
+    n_points=ess_n_points(),
+    extra_methods=ess_extra_methods,
+    min_ess=ess_min_ess(),
```
use `st.xyz` directly here

OriolAbril commented on tests/test_hypothesis_plots.py:
```diff
@@ -192,3 +192,73 @@ def test_plot_ridge(datatree, combined, plot_kwargs, labels_shade_label):
+        {},
+        optional={
+            "ess": plot_kwargs_value,
+            "rug": st.sampled_from(({}, {"color": "red"})),
```
This can be defined outside in line 43 as `plot_kwargs_value_no_false`

OriolAbril commented on tests/test_hypothesis_plots.py:
```diff
@@ -192,3 +192,73 @@ def test_plot_ridge(datatree, combined, plot_kwargs, labels_shade_label):
             assert all(key in child for child in pc.viz.children.values())
+
+
+ess_kind_value = st.sampled_from(("local", "quantile"))
```
also move to line 43

OriolAbril commented on tests/test_plots.py:
```diff
@@ -246,3 +254,51 @@ def test_plot_ridge_aes_labels_shading(self, backend, datatree_4d, pseudo_dim):
+        assert "rug" in pc.viz["mu"]
+        assert "hierarchy" not in pc.viz["mu"].dims
+        assert "hierarchy" in pc.viz["theta"].dims
+        assert pc.viz["mu"].rug.shape == (4,)  # checking rug artist shape (4 chains overlaid)
```
```suggestion
        assert "chain" in pc.viz["mu"].rug.dims
```

OriolAbril commented on tests/test_plots.py:
```diff
@@ -246,3 +254,51 @@ def test_plot_ridge_aes_labels_shading(self, backend, datatree_4d, pseudo_dim):
+        assert "ess" in pc.viz["mu"]
+        assert "min_ess" in pc.viz["mu"]
+        assert "title" in pc.viz["mu"]
+        # assert "rug" in pc.viz["mu"]
```
```suggestion
        assert "rug" not in pc.viz["mu"]
```

OriolAbril commented on tests/test_plots.py:
```diff
@@ -246,3 +254,51 @@ def test_plot_ridge_aes_labels_shading(self, backend, datatree_4d, pseudo_dim):
+        assert "hierarchy" not in pc.viz["mu"].dims
+        assert "hierarchy" in pc.viz["theta"].dims
+        assert "model" in pc.viz["mu"].dims
+        # assert pc.viz["mu"].rug.shape == (2, 4)  # since there are 2 Ms, 4 Cs in datatree
```
delete

OriolAbril reviewed (commented):
very close to merging

imperorrp commented:
Rebased the PR

OriolAbril commented on docs/source/api/plots.rst:
```diff
@@ -23,3 +23,4 @@ A complementary introduction and guide to ``plot_...`` functions is available at
    plot_ridge
    plot_trace
    plot_trace_dist
+   plot_ess
```
can you keep the entries sorted alphabetically?

OriolAbril commented on src/arviz_plots/plots/essplot.py:
```diff
@@ -0,0 +1,585 @@
+"""ess plot code."""
+
+# imports
+# import warnings
```
```suggestion
```

OriolAbril commented on src/arviz_plots/plots/essplot.py:
```diff
@@ -0,0 +1,585 @@
+    if rug_kwargs is False:
+        raise ValueError("plot_kwargs['rug'] can't be False, use rug=False to remove the rug")
+
+    if backend is None:
+        backend = rcParams["plot.backend"]
+    plot_bknd = import_module(f".backend.{backend}", package="arviz_plots")
```
Use same pattern as plot_forest: https://github.com/arviz-devs/arviz-plots/blob/main/src/arviz_plots/plots/forestplot.py#L217-L221

OriolAbril commented on src/arviz_plots/plots/essplot.py:
```diff
@@ -0,0 +1,585 @@
+        backend = rcParams["plot.backend"]
+    plot_bknd = import_module(f".backend.{backend}", package="arviz_plots")
+
+    # set plot collection initialization defaults if it doesnt exist
+
+    # figsizing related plot collection initialization
+    if plot_collection is None:
+        figsize = pc_kwargs.get("plot_grid_kws", {}).get("figsize", None)
+        figsize_units = pc_kwargs.get("plot_grid_kws", {}).get("figsize_units", "inches")
+        pc_kwargs.setdefault("col_wrap", 5)
+        pc_kwargs.setdefault(
+            "cols",
+            ["__variable__"]
+            + [dim for dim in distribution.dims if dim not in {"model"}.union(sample_dims)],
+        )
+        n_plots, _ = process_facet_dims(distribution, pc_kwargs["cols"])
+        col_wrap = pc_kwargs["col_wrap"]
+        if n_plots <= col_wrap:
+            n_rows, n_cols = 1, n_plots
+        else:
+            div_mod = divmod(n_plots, col_wrap)
+            n_rows = div_mod[0] + (div_mod[1] != 0)
+            n_cols = col_wrap
+    else:
+        figsize, figsize_units = plot_bknd.get_figsize(plot_collection)
+        n_rows = leaf_dataset(plot_collection.viz, "row").max().to_array().max().item()
+        n_cols = leaf_dataset(plot_collection.viz, "col").max().to_array().max().item()
+
+    plot_bknd = import_module(f".backend.{backend}", package="arviz_plots")
+
+    figsize = plot_bknd.scale_fig_size(
+        figsize,
+        rows=n_rows,
+        cols=n_cols,
+        figsize_units=figsize_units,
+    )
```
This is the initial version of figsize scaling, it should follow the same pattern as plot_forest. The only thing to take from plot_trace is the computation of number of rows and columns (instead of using the one from plot_forest which would make no sense here).

https://github.com/arviz-devs/arviz-plots/blob/main/src/arviz_plots/plots/forestplot.py#L258-L278, changing https://github.com/arviz-devs/arviz-plots/blob/main/src/arviz_plots/plots/forestplot.py#L271-L272 to the computation plot_trace does to get the number of columns and rows in the generated grid.

imperorrp commented:
Made those modifications. Also, pylint was indicating an error regarding the number of tests under the TestPlots class in testplots.py so I raised the limit from 20 to 25. 

The output is like this right now though: 
![image](https://github.com/user-attachments/assets/a258f29a-4440-4fe7-a15d-192ef4887163)

^All 10 subplots in one row despite the print statements indicating the figsizing logic computes 2 rows and sticks to the 5 max subplots per row constraint. This has to be looked into if I'm missing something.

OriolAbril commented on .pylintrc:
```diff
@@ -474,7 +474,7 @@ max-locals=15
 
 # Maximum number of public methods for a class (see R0904).
-max-public-methods=20
+max-public-methods=25
```
Add a comment in the `test_plots` file instead. https://github.com/arviz-devs/arviz-plots/blob/main/tests/test_plots.py#L1

OriolAbril commented on src/arviz_plots/plots/essplot.py:
```diff
@@ -0,0 +1,589 @@
+            figsize_units = "dots"
+        pc_kwargs["plot_grid_kws"]["figsize"] = figsize
+        pc_kwargs["plot_grid_kws"]["figsize_units"] = figsize_units
+        plot_collection = PlotCollection.grid(
```
You are using `.grid` instead of `.wrap` so the grid no longer "wraps" the 1sr row into multiple ones when it has too many columns.

imperorrp replied:
Yeah using `.wrap` worked: 

![image](https://github.com/user-attachments/assets/14d4201c-a508-4fdc-9101-e25f85c89843)

OriolAbril commented on src/arviz_plots/plots/essplot.py:
```diff
@@ -0,0 +1,589 @@
+        figsize = pc_kwargs.get("plot_grid_kws", {}).get("figsize", None)
+        figsize_units = pc_kwargs.get("plot_grid_kws", {}).get("figsize_units", "inches")
+        if figsize is None:
+            coeff = 0.2
```
All the coeff related part was plot_forest specific, here the variable isn't used anywhere

imperorrp commented:
Todo: Add `plot_ess_evolution`, `plot_mcse` reference in 'see also' once both this and their PRs (#71, #79) are merged.
